ocrd_calamari/recognize.py

```diff
@@ -54,17 +62,30 @@ def process(self):
             pcgts = page_from_file(self.workspace.download_file(input_file))
 
             page = pcgts.get_Page()
-            page_image, page_xywh, page_image_info = self.workspace.image_from_page(page, page_id)
+            page_image, page_coords, page_image_info = self.workspace.image_from_page(
+                page, page_id, feature_selector=self.features)
 
-            for region in pcgts.get_Page().get_TextRegion():
-                region_image, region_xywh = self.workspace.image_from_segment(region, page_image, page_xywh)
+            for region in page.get_TextRegion():
+                region_image, region_coords = self.workspace.image_from_segment(
+                    region, page_image, page_coords, feature_selector=self.features)
 
                 textlines = region.get_TextLine()
                 log.info("About to recognize %i lines of region '%s'", len(textlines), region.id)
-                for (line_no, line) in enumerate(textlines):
-                    log.debug("Recognizing line '%s' in region '%s'", line_no, region.id)
-
-                    line_image, line_xywh = self.workspace.image_from_segment(line, region_image, region_xywh)
+                for line in textlines:
+                    log.debug("Recognizing line '%s' in region '%s'", line.id, region.id)
+
+                    line_image, line_coords = self.workspace.image_from_segment(
+                        line, region_image, region_coords, feature_selector=self.features)
+                    if ('binarized' not in line_coords['features'] and
+                        'grayscale_normalized' not in line_coords['features'] and
+                        self.input_channels == 1):
+                        # We cannot use a feature selector for this since we don't
+                        # know whether the model expects (has been trained on)
+                        # binarized or grayscale images; but raw images are likely
+                        # always inadequate:
+                        log.warning("Using raw image for line '%s' in region '%s'",
```

Comment:
Is "raw image" the proper term? I wouldn't understand what the problem is with this message.

Reply:
Also, is there a standard for line length in the OCR-D project?

Reply:
> Is "raw image" the proper term? I wouldn't understand what the problem is with this message.

I think it is. This gives the correct warning for the current CI tests: the images are raw colors, not greyscale, not normalized, not binarized.

Reply:
> Also, is there a standard for line length in the OCR-D project?

I wouldn't know. Why? Too long, or too short for your taste?

Reply:
> Also, is there a standard for line length in the OCR-D project?

Generally PEP8 so 72. But I've disabled this in my pylint setups because it's just too restrictive. So, no limit :)

Reply:
> "raw image"

"Using raw image without any preprocessing ..." ?

Reply:
> "Using raw image without any preprocessing ..." ?

You don't know that. It could have been cropped, deskewed, dewarped – just not binarized or grayscale-normalized (either of which is necessary for a network with 1 input channel).

`Raw` is the best (and shortest) unique description of _RGB or unnormalized greyscale_ I know.